zerver/openapi/zulip.yaml

```diff
@@ -1222,6 +1226,13 @@ paths:
                             - type: object
                               description: |
                                 Event sent to all users who can see a stream when it is deactivated.
+
+                                This event is also sent when a user loses access to a stream they
+                                previously [could access](/help/stream-permissions) when their roles
+                                changes.
+
+                                **Changes**: Prior to Zulip 8.0 (feature level 205), this event was
+                                not sent when a user lost access to a stream due to role change.
```

Comment:
@laurynmm can I ask you to do a follow-up editing pass on these descriptions? I think mostly editing for clarity and also trying to make clear that the feature level 205 change is that the user should now get this event when they stop being able to access a stream due to permissions change such as a role change -- I want it to be the case that if removing a user from a group or some other event causes them to lose access to the stream, it's clear that clients understand they should expect to be notified this way too.

Might be worth an api design discussion if I'm confused here, but @sahil839 am I correct that it was previously possible to get this event when a user was removed from a private stream too?

Reply:
@timabbott No, we do not sent this event when user is unsubscribed from a private stream, because the users (irrespective of role) have access to the unsubscribed private streams, but not the private streams they have never been subscribed too. And similar for guests as well, they get the data for unsubscribed public and private streams.

Reply:
So, when a user's role changes and they can no longer access a stream, they do still have access to the messages in the stream that they sent/read in those streams, correct?

Reply:
By having access to a stream, I mean that user at least knows that such a stream exists and may or may not have other permissions like subscribing to stream or access to its messages.

So, there are two cases where user does not have access to a stream -
- For guests, they do not have access to a public stream that they have never been subscribed too. But they would have access to unsubscribed public streams (by unsubscribed I mean they were subscribed at some point of time but are not subscribed currently). 
- For users who are not admins or owners, they do not have access to a private stream that they have never been subscribed to, but they would have access to unsubscribed private streams.

And I think in the above cases, the user would not have access to any of the messages in those streams.

Reply:
So, when a user is unsubscribed (whether themselves or by someone else), they do not lose access to a stream. And we follow the same logic when sending events for role changes. Which means that we do not send "stream op: delete" event for unsubscribed public or private streams when a user is changed from a member to guest or from an admin to member/moderator/guest.

Reply:
> Which means that we do not send "stream op: delete" event for unsubscribed public or private streams when a user is changed from a member to guest or from an admin to member/moderator/guest.

So, when testing in the dev environment, when I change a user from member to guest, the `streams op: delete` event sent includes data for all currently unsubscribed public streams. It doesn't seem to matter if the user has never been subscribed or previously was subscribed. Maybe we should move to a CZO conversation?

Also, when I change a user from guest to member, a `subscriptions op: peer_add` event is sent for each stream in the `streams op: add` event, which we didn't document in the feature level 205 as a change. So I think we need to add that, correct?

Reply:
Yes, I caught that bug involving guest users today while working on another issue and I intend to start a discussion on CZO today. 

> Also, when I change a user from guest to member, a subscriptions op: peer_add event is sent for each stream in the streams op: add event, which we didn't document in the feature level 205 as a change. So I think we need to add that, correct?

Yes, we would need to add that.

Reply:
I'll draft up the feature level update in a PR and flag you for review when it's ready.